src/dynamixel_hardware_interface.cpp

```diff
@@ -51,6 +52,22 @@ namespace dynamixel {
             throw e;
         }
 
+        // remove servos that are not in the _dynamixel_map (i.e. that are not used)
+        std::vector<dynamixel::DynamixelHardwareInterface::dynamixel_servo>::iterator servo_it;
+        for (servo_it = _dynamixel_servos.begin(); servo_it != _dynamixel_servos.end();) {
+            std::map<long long int, std::string>::iterator dynamixel_iterator = _dynamixel_map.find((*servo_it)->id());
+            if (dynamixel_iterator == _dynamixel_map.end()) // the actuator's name is not in the map
+                servo_it = _dynamixel_servos.erase(servo_it);
+            else
+                ++servo_it;
+        }
+
+        // Check that no actuator was declared by user but not found
+        if (_dynamixel_map.size() > _dynamixel_servos.size()) {
+            ROS_FATAL_STREAM("Some servos were declared in the parameters but could "
+                             "not be found");
+        }
```

Comment:
During my tests, this case always ended in the controller manager 
complaining and crashing because the resource asked by a controller was 
not available.

Needs more testing...

Le 17/03/2016 12:48, Konstantinos Chatzilygeroudis a écrit :
>
> In src/dynamixel_hardware_interface.cpp 
> <https://github.com/resibots/dynamixel_control_hw/pull/4#discussion_r56491531>:
>
> > @@ -51,6 +52,22 @@ namespace dynamixel {
> >              throw e;
> >          }
> >
> > +        // remove servos that are not in the _dynamixel_map (i.e. that are not used)
> > +        std::vector<dynamixel::DynamixelHardwareInterface::dynamixel_servo>::iterator servo_it;
> > +        for (servo_it = _dynamixel_servos.begin(); servo_it != _dynamixel_servos.end();) {
> > +            std::map<long long int, std::string>::iterator dynamixel_iterator = _dynamixel_map.find((*servo_it)->id());
> > +            if (dynamixel_iterator == _dynamixel_map.end()) // the actuator's name is not in the map
> > +                servo_it = _dynamixel_servos.erase(servo_it);
> > +            else
> > +                ++servo_it;
> > +        }
> > +
> > +        // Check that no actuator was declared by user but not found
> > +        if (_dynamixel_map.size() > _dynamixel_servos.size()) {
> > +            ROS_FATAL_STREAM("Some servos were declared in the parameters but could "
> > +                             "not be found");
> > +        }
>
> We should not output a fatal error in this case. Just a warning is 
> enough. For example, when we are using the broken robots we are always 
> declaring actuators that are not there in the end.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/resibots/dynamixel_control_hw/pull/4/files/034387d41db6779c19b0fed9e676fcb0851a086e#r56491531>
>